Add history dock to default editor layout, and prevent signal connecting multiple times

souplamp opened the pull request:
Fixes: #69069, possibly #68882

### **Changes:**
- Change enter tree notification to ready notification to ensure the `HistoryDock::on_history_changed` signal only attempts to connect once (as the enter tree notification is fired whenever the dock is moved in the editor layout, see discussion).
- Add the history dock to the default editor layout; using Editor -> Editor Layout -> Default will no longer make the history dock disappear.

KoBeWi commented:
I can't reproduce the issue.

Also you could just change `NOTIFICATION_ENTER_TREE` to `NOTIFICATION_READY`.

souplamp commented:
> I can't reproduce the issue.

I'm not entirely sure why it's happening, and I was the first to open an issue on it so it seems to be just me. Unless History Dock is somehow getting the Enter Tree notification twice?

> Also you could just change `NOTIFICATION_ENTER_TREE` to `NOTIFICATION_READY`.

That is much simpler than my implementation, thank you. Updated.

KoBeWi commented:
There is no way the NOTIFICATION_ENTER_TREE could be received twice, unless the dock somehow leaves the scene tree. Do you use some custom editor layout? Does the issue happen in a fresh project?

akien-mga commented:
Does the dock leave the scene tree temporarily when moved to a new dock slot maybe?

KoBeWi commented:
Yeah but most of the editor code uses enter tree instead of ready, so not sure why it would be a problem here.

souplamp commented:
> There is no way the NOTIFICATION_ENTER_TREE could be received twice, unless the dock somehow leaves the scene tree. Do you use some custom editor layout? Does the issue happen in a fresh project?

You're right, it was the custom editor layout. I place all my docks on the right-side of the screen. Can't reproduce with a fresh project and default layout.

KoBeWi commented:
Is this a global editor layout? I'm not familiar with this feature, you might be experiencing issue similar to https://github.com/godotengine/godot/issues/68882#issuecomment-1324289021. I think it should be gone after you resave your layout with History dock (which would make this a temporary version transition issue).

souplamp commented:
> Is this a global editor layout? I think it should be gone after you resave your layout with History dock (which would make this a temporary version transition issue).

Saving the history dock position does not fix the error unfortunately. I tested in 4.0beta6 as well, just to be sure, but I still am getting the issue.

souplamp committed:
History dock singleton, set default editor layout, ready notification

- add the history dock to the default editor layout, so when a user does Editor -> Editor Layout -> Default the history dock will no longer disappear
- change the enter tree notification to a ready notification to prevent the history dock from trying to connect 'on_history_changed' signal everytime the dock is moved in the editor layout

souplamp commented:
I did some investigating and figured out what's happening.
Essentially, akien was right in saying this should be a ready notification, not an enter tree notification.

> Most of the editor code uses enter tree instead of ready, so not sure why it would be a problem here.

This is true, but the other editor docks handle the case where the dock has already been initialized.

Take [filesystem_dock.cpp](https://github.com/godotengine/godot/blob/master/editor/filesystem_dock.cpp#L349) for example:

```C++
case NOTIFICATION_ENTER_TREE: {
	if (initialized) {
		return;
	}
	initialized = true;
```

This check is done because Enter Tree happens every time a dock is moved in the editor layout.

For example, we add the following to [node_dock.cpp](https://github.com/godotengine/godot/blob/master/editor/node_dock.cpp#L56):

```C++
case NOTIFICATION_ENTER_TREE: {
	WARN_PRINT("I am the Node Dock!! Fear me!!");
} break;
```

We get this:
![node dock](https://user-images.githubusercontent.com/89402588/203682210-e264c3ff-b00c-4f64-8af8-cc26498e5ea2.gif)

The Enter Tree notification fires whenever the dock is moved in the editor layout, thus connecting the signal every time the history dock is moved.

We get the same node dock test print whenever we load a new project:
![tst](https://user-images.githubusercontent.com/89402588/203683975-9073007e-0b79-4b14-9299-c114988ac8e9.png)

Notice something missing? Since we're using a fresh project with the global editor layout, the history dock enters the tree once. But if we load a previous project with a custom layout, the editor (only my hunch) will load the global default layout, _then_ load the custom editor layout.

This second loading of the editor layout will trigger the enter tree notification in [history_dock.cpp](https://github.com/godotengine/godot/blob/master/editor/history_dock.cpp#L209) and attempt to connect the signal a second time, causing the error.

So in all, it is the right move to make this a ready notification. But my curiosity got the best of me, so I wanted to figure out why :)

KoBeWi commented:
I don't see the point in making HistoryDock a singleton if it's not used outside EditorNode >_>

souplamp commented:
> I don't see the point in making HistoryDock a singleton if it's not used outside EditorNode >_>

There's no use case at the moment, so it can be removed from this PR if necessary. Mainly added for consistency, and possible future-proofing for something like EditorPlugin.

KoBeWi commented:
I'd leave it non-singleton for now. It's easy enough to change if it's ever needed.

souplamp commented:
Done.

akien-mga commented:
Thanks!
